rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,406 @@
+      tls:
+        algorithm: ECDSA_P256_SHA256 # always use this for TLS keys
+      jwt:
+        algorithm: recommended
+```
+
+```yaml
+# teleport.yaml
+version: v3
+auth_service:
+  enabled: true
+  ca_key_params:
+    gcp_kms:
+      keyring: projects/teleport-dev-320620/locations/us-west1/keyRings/nic-testplan-13
+      protection_level: "SOFTWARE"
+    default:
+      ssh:
+        algorithm: recommended
+      tls:
+        algorithm: recommended
+      jwt:
+        algorithm: recommended
+    host:
+      ssh:
+        algorithm: ECDSA_P256_SHA256
+      tls:
+        algorithm: ECDSA_P256_SHA256
+    user:
+      ssh:
+        algorithm: Ed25519
+      tls:
+        algorithm: ECDSA_P256_SHA256
+    db:
+      tls:
+        algorithm: RSA2048
+    jwt:
+      jwt:
+        algorithm: recommended # let Teleport choose
+```
+
+#### Cloud
+
+Cloud will be able to select their preferred defaults by configuring them in the
+`teleport.yaml`
+
+Should CA algorithms be configurable by Cloud users?
+I don't see why not, we will only support secure algorithms.
+They can choose their own algorithms be editing the `cluster_auth_preference`
+and doing a CA rotation.
+
+### Subjects
+
+It will be nice to update the CA key algorithms for security and performance
+benefits, but what users really see on a day-to-day basis is their user keys.
+
+"Subjects" that have certificates issued by the Teleport CAs include:
+
+* Teleport users via `tsh login`
+* Teleport users via `tsh app login`
+* Teleport users via `tsh db login`
+* Teleport users via Teleport Connect
+* Teleport services (ssh, app, db, kube, windows desktop, etc)
+* Machine ID (`tbot`)
+* Teleport Plugins
+* OpenSSH hosts
+
+All of these currently generate an RSA2048 keypair locally, send the public key
+to the auth server, and receive signed certificates of some variety.
+
+We have a few options for how to evolve this:
+
+1. Choose a new unconfigurable algorithm to use for each client.
+2. Try to match the algorithm of the relevant CA for each.
+3. Make each configurable cluster-wide. (this is similar to (2) but doesn't
+   require subject and CA algorithms to match).
+4. Make each configurable by each subject (like `tsh login --keytype ecdsa_P256`, new field in `teleport.yaml` for hosts, etc).
+
+My personal opinion: The MVP is (2).
+(1) is not flexible enough, (2) seems like it would cover >80% of cases, (3)
+adds a few more knobs without major benefits, someone will eventually ask for
+(4) and we can consider adding it then.
+
+### Backward Compatibility
+
+* Will the change impact older clients? (tsh, tctl)
+
+Auth servers with non-default algorithms configured should continue to sign
+certificates for clients on older Teleport versions using RSA2048 keys.
+
+Open questions:
+
+* Should we start rejecting RSA2048 (if a different algo is configured) in
+  future Teleport versions (V15)?
+* Should we make some configurable `allowed_subject_algorithms` per CA to put
```

Comment:
It's unclear to me if there is a case for supporting multiple algorithms.  The security value will only be based off the weakest algorithm supported.  Is there a case where a user would benefit from having multiple algorithms?

Reply:
The purpose of this would be compatibility, not added security. Perhaps we introduce this feature in 14.0.0 but the user still has hosts on 13.1.2 that only know how to generate RSA keys. By default I believe we should allow this and sign their certificate so that we can keep compatibility. Also, maybe we want to give the option to generate OpenSSH host certs or user certs with a different algorithm that is supported by their version or OpenSSH or a different software. But if the Teleport admin wants to deny the use of RSA everywhere, they could remove it from this `allowed_subject_algorithms` list